### DISCUSSION TITLE
Native data package management for Android

### DISCUSSION THREAD

**Discussion by Olga-Yakovleva** [RHVoice/RHVoice#453] [Show and tell]
Hi everyone,

I've implemented support for getting the information about the language and voice packages from the internet. See #452. The Android app has been updated to use this feature. It's a lot of changes. I need Alpha testers for the app. The first task is to make sure that upgrading from the current stable version of the app runs smootthly. Than that fresh installs work correctly. After that we'll start testing an update of a data package. Check that no other existing functionality got broken. Of course, I want to know if this all works when using RHVoice with Talkback all the time. So, please, if anyone wants to help or can invite someone who wants to help, send me a request with the gmail address to rhvoice@rhvoice.org. But please don't apply if you are just curious to see what's new. I make no stability promises about this new code.

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
I've applied to help you with testing. Also, the compilation with scons is failing
i get the following error:
In file included from external/libs/boost/include/boost/mpl/or.hpp:43,                                                  
                 from external/libs/boost/include/boost/iterator/interoperable.hpp:11,                                  
                 from external/libs/boost/include/boost/iterator/iterator_facade.hpp:11,                                
                 from external/libs/boost/include/boost/range/iterator_range_core.hpp:27,                               
                 from external/libs/boost/include/boost/range/iterator_range.hpp:13,                                    
                 from external/libs/boost/include/boost/range/as_literal.hpp:18,                                        
                 from external/libs/boost/include/boost/algorithm/string/trim.hpp:19,                                   
                 from external/libs/boost/include/boost/algorithm/string.hpp:19,                                        
                 from src/core/voice.cpp:18:                                                                            
external/libs/boost/include/boost/mpl/aux_/include_preprocessed.hpp:37:89: fatal error: boost/mpl/aux_/preprocessed/gcc/or.hpp: No such file or directory                                                                                       
   37 | #   include BOOST_PP_STRINGIZE(boost/mpl/aux_/preprocessed/AUX778076_PREPROCESSED_HEADER)                       
      |                                                                                         ^                       
compilation terminated.                                                                                                 
scons: *** [build/linux/core/voice.os] Error 1                                                                          
scons: building terminated because of errors.

**Reply by Olga-Yakovleva** [reply]
I've added you and sent you a notification message with further instructions. And thank you for the bug report.

**Comment by zstanecic** [comment]
Great. I've accepted the invitation, and the upgrade process ran smoothly.
How to use new approach to add voices and languages to the android app?
How the app downloads json from the internet?

**Comment by Olga-Yakovleva** [comment]
Please see here for details and instructions for language and voice 
contributors: https://github.com/RHVoice/packages/

**Comment by zstanecic** [comment]
Great. The bug with compiling is fixed. on the dpm branch. 
Now the second question is: how to get the base64 checksum of the respective voice data files? Which tool to use on linux / windows?

**Reply by Olga-Yakovleva** [reply]
@zstanecic 
openssl md5 -binary <file>|base64

**Comment by grzezlo** [comment]
Just curious, why so non-standard MD5 notations are used by RHVoice?
Earlier an arrays of unsigned bytes, now something looking like base64 encoding of binary value...
Why not old, well-known and widely used hexadecimal MD5 notation, easy to calculate using many tools for Windows, or md5sum under Linux?
$ md5sum somefile.zip
963f33b5232e772030d3a6c3963149f4

**Reply by Olga-Yakovleva** [reply]
@grzezlo 
Previously I used arrays of bytes I could put directly into Java source files, because all the information was coded as Java classes.

Now it's base64, because base64 seems to be the standard way to encode binary data when getting it from your server if your data format is json. And I don't need to create any hand-written little parsers for it, which is not the case with the hex encoding, for whitch I couldn't find any builtin API to convert it back to binary on Android.

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
This command gives multiple output. Which to use? i.e, it looks like that i get multiple sums for one file wrapped into multiple lines

**Reply by Olga-Yakovleva** [reply]
It gives just one line here on my machine. Something is going wrong on yours, but I'm not sure what it might be.

**Comment by zstanecic** [comment]
I am running the command as follows:
openssl md5 -binary |base64 filename
is it wrong?

**Comment by Olga-Yakovleva** [comment]
@zstanecic, @beqabeqa473, 

I have a bug report from a tester:

> After its update, my phone went to using Samsung TTS voice for US English, even though RH Voice was selected in the prefered text to speech engine setting. So I restarted the phone, after the restart, again Samsung’s voice was there  instead of RH Voice

I need to know if this can be reproduced. I myself and another tester haven't been able to reproduce. But this Alpha isn't going to production until I can figure out if this bug is in RHVoice itself and what's going on if it is. So, could you help by doing extensive testing of the following?

1. Uninstall the alpha and install the production version either by temporarily leaving the testing program or sideloading the app using the direct link: https://rhvoice.org/download/RHVoice-v1.6.1.apk. If you sideload, make sure that Android doesn't update the app before we need it to, for example, turn your network connection off for a time.

2. Set the production version up and start using it with TalkBack.

3. If necessary, rejoin the testing program using the same link.

4. Update RHVoice while still using it with TalkBack.

5. See what happens. Expected behaviour is TalkBack keeps using RHVoice with a possible silence of a few seconds while the app is being updated and restarted.

And I also need more people testing this. I don't like the idea of asking general user community to test something which may be unstable yet. So, any advanced testers people here can bring will be welcome!

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
Before trying to reproduce the problem, can you verify that the user is not using screen lock method, like pin or fingerprint?
Samsung tts works 5 seconds after the screen unlock then RHVoice starts talking normally.

**Comment by zstanecic** [comment]
Ps, when the stable version is updating to beta, RHVoice turns off, until you manually run the RHvoice from it's shortcut on the home screen.

**Reply by Olga-Yakovleva** [reply]
Do you mean this is what is happening on your phone?

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
yes, of course.

**Comment by zstanecic** [comment]
It is on android 11, samsung galaxy a71.

**Comment by Olga-Yakovleva** [comment]
This sounds like the same thing as in the other bug report to me. It's not what should be happening, but you haven't reported it before. On my phone RHVoice restarts and keeps speaking.

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
I haven't reported it, because i didn't knew, that it shouldn't happen. 
I don't even know, how can i give you any log of it. As i said, i worked this around by clicking the rhvoice icon on the home screen, as it clises after the update and switches to the samsung tts, but it runs again when you turn it on, and even it works after the phone restart.

**Comment by Olga-Yakovleva** [comment]
Two samsung phones where it is happening then. One Nokia phone where it isn't. More data is needed.

**Comment by Olga-Yakovleva** [comment]
As for logs, right now I don't do any logging in those builds. Android doesn't make logging simple in the builds uploaded to Google Play.

**Comment by zstanecic** [comment]
@Olga-Yakovleva 
which extra data you wil need? what can i do more to debug this?

**Reply by Olga-Yakovleva** [reply]
I meant more testers. At least to figure out if it's general or Samsung-specific. Though it would be helpful if you could see if this behaviour is new to the alpha version. Can you sideload apps on your phone, not just install from Google Play? Then we could see if updating from one older production version to another causes the same situation.

**Comment by zstanecic** [comment]
Where i can find the apk to the beta?

**Comment by Olga-Yakovleva** [comment]
No, we don't want the test version for this. If we are trying to figure 
out if the older stable versions did the same. So please uninstall the 
beta, leave the testing program for now using the same link you used to 
join, then install this previous production version, 1.6.0:

https://rhvoice.org/download/RHVoice-v1.6.0.apk

Set it up with Talkback, then update to the current production version, 
1.6.1, through Google Play.

**Comment by JonnyTech** [comment]
I am interested in becoming a tester, but use [F-Droid](https://f-droid.org/en/packages/com.github.olga_yakovleva.rhvoice.android), can I participate?

**Comment by zstanecic** [comment]
Hi
@Olga-Yakovleva 
Again,
After the upgrade to 1.6.1 from sideloaded 1.6.0, synthesizer restarted automatically after upgrade immediatelly.

**Comment by zstanecic** [comment]
Yet another update: this problem disappeared after updating to the 1.7.9, which is very strange. This can be a very random problem to catch.

**Comment by beqabeqa473** [comment]
I am not able to update to alpha.

Unfortunately, i am not able to test changes.

Could you please upload somewhere apk for latest version?
I could build it myself but i won't be able to test problem you described.

For an information, samsung phones, maybe android 12 itself can cause
this type ot troubles, in this case it is necesary to restart
accessibility service such as talkback.
Directboot support would solve this problem once at all.

On 1/29/22, Zvonimir Stanečić ***@***.***> wrote:
> Yet another update: this problem disappeared after updating to the 1.7.9,
> which is very strange. This can be a very random problem to catch.
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/RHVoice/RHVoice/discussions/453#discussioncomment-2070750
> You are receiving this because you were mentioned.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by Olga-Yakovleva** [comment]
Here's the direct link to the apk:

https://rhvoice.org/download/RHVoice-1.7.9.apk

**Comment by Olga-Yakovleva** [comment]
@zstanecic, 

What was the user interface language of your phone during the update which stopped speaking? And which RHVoice voice was running at the time?

**Comment by zstanecic** [comment]
Interface language: russian,

Voice, mikhail +kiko

**Comment by beqabeqa473** [comment]
cannot confirm a problem here on samsung running android 12, and
Xiaomi phone running android 10.

On 1/31/22, Zvonimir Stanečić ***@***.***> wrote:
> Interface language: russian,
>
> Voice, mikhail +kiko
>
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/RHVoice/RHVoice/discussions/453#discussioncomment-2083486
> You are receiving this because you were mentioned.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by zstanecic** [comment]
As i said, 

This problem disappeared after the sideloaded updates from one version to another, then joining the beta program. I don’t know what can cause the problem.
